lgc/builder/llpcYCbCrAddressHandler.cpp

```diff
@@ -0,0 +1,196 @@
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+***********************************************************************************************************************
+* @file  llpcYCbCrAddressHandler.cpp
+* @brief LLPC source file: Implementation of LLPC class YCbCrAddressHandler
+***********************************************************************************************************************
+*/
+#include "llpcYCbCrAddressHandler.h"
+#include "llpcGfxRegHandler.h"
+#include "llpcInternal.h"
+
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+
+using namespace lgc;
+using namespace llvm;
+
+// =====================================================================================================================
+// The class is used for calculating and mantain the base address of each plane in YCbCr image.
+void YCbCrAddressHandler::GenBaseAddress(
+    uint32_t planeNum) // The number of planes
+{
+    // For YCbCr, the possible plane numbers are among 1 and 3.
+    assert(planeNum > 0 && planeNum < 4);
+
+    if (planeNum > 0)
+    {
+        // m_PlaneBaseAddresses[0] is the same as original base addressed as passed in SRD
+        m_PlaneBaseAddresses.push_back(m_pRegHelper->GetReg(SqRsrcRegs::BaseAddress));
+        if (planeNum > 1)
+        {
+            // m_PlaneBaseAddresses[1] = m_PlaneBaseAddresses[0] + addr_256B(m_pPitchY * m_pHeightY)
+            m_PlaneBaseAddresses.push_back(m_pBuilder->CreateAdd(m_PlaneBaseAddresses[0],
+                                           m_pBuilder->CreateLShr(m_pBuilder->CreateMul(m_pPitchY, m_pHeightY),
+                                           m_pBuilder->getInt32(8))));
+            if (planeNum > 2)
+            {
+                // m_PlaneBaseAddresses[2] = m_PlaneBaseAddresses[1] + addr_256B(m_pPitchCb * m_pHeightCb)
+                m_PlaneBaseAddresses.push_back(m_pBuilder->CreateAdd(m_PlaneBaseAddresses[1],
+                                               m_pBuilder->CreateLShr(m_pBuilder->CreateMul(m_pPitchCb, m_pHeightCb),
+                                               m_pBuilder->getInt32(8))));
+            }
+        }
+    }
+}
+
+// =====================================================================================================================
+// Power2align operation
+Value* YCbCrAddressHandler::Power2Align(
+    Value*   pX,    // [in] Value needs to be aligned
+    uint32_t align) // Align base
+{
+    // Check if align is power2 aligned
+    assert(align != 0 && (align & (align-1)) == 0);
+
+    uint32_t alignM1 = align - 1;
+    uint32_t alignM1Inv = ~alignM1;
```

Comment:
I think you don't need to define alignM1 and alignM1Inv, use `m_pBuilder->getInt32(~(align - 1))`
If there are type mismatch, use static_cast<uint32_t>().